spirv_msl.cpp

```diff
@@ -320,12 +379,31 @@ static string create_sampler_address(const char *prefix, MSLSamplerAddress addr)
 	}
 }
 
+SPIRType &CompilerMSL::get_stage_in_struct_type()
+{
+	auto &si_var = get<SPIRVariable>(stage_in_var_id);
+	return get_variable_data_type(si_var);
+}
+
 SPIRType &CompilerMSL::get_stage_out_struct_type()
 {
 	auto &so_var = get<SPIRVariable>(stage_out_var_id);
 	return get_variable_data_type(so_var);
 }
 
+SPIRType &CompilerMSL::get_patch_stage_out_struct_type()
+{
+	auto &so_var = get<SPIRVariable>(patch_stage_out_var_id);
+	return get_variable_data_type(so_var);
+}
+
+std::string CompilerMSL::get_tess_factor_struct_name()
+{
+	if (get_entry_point().flags.get(ExecutionModeTriangles))
```

Comment:
Is this known in the control shader? IIRC, this is only known in the evaluation shader.

Reply:
What about line tessellation?

Reply:
> Is this known in the control shader? IIRC, this is only known in the evaluation shader.

The Vulkan spec says that either one can have it set. The intent is that, if it's set on the eval shader, MoltenVK will copy this to the control shader. I don't know what to do about offline compilation.

> What about line tessellation?

That's actually unsupported by Metal.

Reply:
Hm, right. I guess worst case we just need some offline override options, but this is niche enough. Could we have an error if line tessellation is used at least?

Reply:
Done.